private/pkg/git/repository.go

```diff
@@ -153,8 +153,66 @@ func (r *repository) DefaultBranch() string {
 	return r.defaultBranch
 }
 
-func (r *repository) CurrentBranch(ctx context.Context) (string, error) {
-	return detectCheckedOutBranch(ctx, r.gitDirPath, r.runner)
+func (r *repository) CheckedOutBranch(options ...CheckedOutBranchOption) (string, error) {
+	var config checkedOutBranchOpts
+	for _, option := range options {
+		option(&config)
+	}
+	headBytes, err := os.ReadFile(filepath.Join(r.gitDirPath, "HEAD"))
+	if err != nil {
+		return "", fmt.Errorf("read HEAD bytes: %w", err)
+	}
+	headBytes = bytes.TrimSuffix(headBytes, []byte{'\n'})
+	// .git/HEAD could point to a named ref, or could be in a dettached state, ie pointing to a git
+	// hash. Possible values:
+	//
+	// "ref: refs/heads/somelocalbranch"
+	// "ref: refs/remotes/someremote/somebranch"
+	// "somegithash"
+	const refPrefix = "ref: refs/"
+	if strings.HasPrefix(string(headBytes), refPrefix) {
+		refRelDir := strings.TrimPrefix(string(headBytes), refPrefix)
+		if config.remote == "" {
+			// only match local branches
+			const localBranchPrefix = "heads/"
+			if !strings.HasPrefix(refRelDir, localBranchPrefix) {
+				return "", fmt.Errorf("git HEAD %s is not pointing to a local branch", string(headBytes))
+			}
+			return strings.TrimPrefix(refRelDir, localBranchPrefix), nil
+		}
+		// only match branches from the specific remote
+		remoteBranchPrefix := "remotes/" + config.remote + "/"
+		if !strings.HasPrefix(refRelDir, remoteBranchPrefix) {
+			return "", fmt.Errorf("git HEAD %s is not pointing to branch in remote %s", string(headBytes), config.remote)
+		}
+		return strings.TrimPrefix(refRelDir, remoteBranchPrefix), nil
+	}
+	// if HEAD is not a named ref, it could be a dettached HEAD, ie a git hash
+	headHash, err := parseHashFromHex(string(headBytes))
+	if err != nil {
+		return "", fmt.Errorf(".git/HEAD is not a named ref nor a git hash: %w", err)
+	}
+	// we can compare that hash with all repo branches' heads
+	var (
+		currentBranch      string
+		foundBranchStopErr = errors.New("found branch, stop looping branches")
+	)
+	if err := r.ForEachBranch(
+		func(branch string, branchHEAD Hash) error {
+			if headHash == branchHEAD {
+				currentBranch = branch
+				return foundBranchStopErr
+			}
+			return nil
+		},
+		ForEachBranchWithRemote(config.remote),
+	); err != nil {
+		if errors.Is(err, foundBranchStopErr) {
+			return currentBranch, nil
+		}
+		return "", fmt.Errorf("for each branch: %w", err)
+	}
+	return "", errors.New("git HEAD is detached, no matches with any branch")
```

Comment:
Can we switch this around a little?
```suggestion
	); err != nil && !errors.Is(err, foundBranchStopErr) {
		return "", fmt.Errorf("for each branch: %w", err)
	}
	if currentBranch == "" {
		return "", errors.New("git HEAD is detached, no matches with any branch")
	}
	return currentBranch, nil
```